Add an option to run RPC tests with TCP init

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48248 Add an option to run RPC tests with TCP init**

We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds an environment variable `RPC_INIT_WITH_TCP` that allows us to run any RPC test with TCP initialization. 

To avoid port collisions, we use `common.find_free_port()`.
Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

rohan-varma committed:
[WIP] Run RPC tests with TCP init in addition to file

We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds testing with TCP init to the rpc tests. We disable/modify a few tests that would be broken under TCP initialization and will fix them in follow up PRs.

Running the tests with TCP init can be controlled with the env var setting `RPC_INIT_WITH_TCP="1"`. In OSS, we add a custom handler to run tests with the file init and TCP init in CI. OSS tests can be run with [todo]. See internal test plan for internal details on testing with TCP init.

To avoid port collisions, we use `common.find_free_port()`.

Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Run RPC tests with TCP init in addition to file"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds testing with TCP init to the rpc tests. We disable/modify a few tests that would be broken under TCP initialization and will fix them in follow up PRs.

Running the tests with TCP init can be controlled with the env var setting `RPC_INIT_WITH_TCP="1"`. In OSS, we add a custom handler to run tests with the file init and TCP init in CI. OSS tests can be run with `python test/run_test.py --verbose -i distributed/rpc/test_process_group_agent`. See internal test plan for internal details on testing with TCP init.

To avoid port collisions, we use `common.find_free_port()`.

Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc_utils.py:
```diff
@@ -37,6 +42,20 @@
+    use_tcp_init = os.environ.get("RPC_INIT_WITH_TCP", None)
+    if use_tcp_init== "1":
+        os.environ["MASTER_ADDR"] = '127.0.0.1'
+        os.environ["MASTER_PORT"] = str(find_free_port())
```
As I understand it, this actually may not be enough to prevent port collisions if there are multiple tests running in parallel? Test 1 could get port X, but before it creates the store and binds the port, Test 2 could also get port X and we would still have the race. If my understanding is correct, we probably need to write currently in use ports to a shared/locked data structure to completely avoid races?

rohan-varma replied:
(This probably won't be an issue anymore now that we are not enabling it for all tests)

rohan-varma committed:
Update on "[WIP] Run RPC tests with TCP init in addition to file"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds testing with TCP init to the rpc tests. We disable/modify a few tests that would be broken under TCP initialization and will fix them in follow up PRs.

Running the tests with TCP init can be controlled with the env var setting `RPC_INIT_WITH_TCP="1"`. In OSS, we add a custom handler to run tests with the file init and TCP init in CI. OSS tests can be run with `python test/run_test.py --verbose -i distributed/rpc/test_process_group_agent`. See internal test plan for internal details on testing with TCP init.

To avoid port collisions, we use `common.find_free_port()`.

Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Run RPC tests with TCP init in addition to file"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds testing with TCP init to the rpc tests. We disable/modify a few tests that would be broken under TCP initialization and will fix them in follow up PRs.

Running the tests with TCP init can be controlled with the env var setting `RPC_INIT_WITH_TCP="1"`. In OSS, we add a custom handler to run tests with the file init and TCP init in CI. OSS tests can be run with `python test/run_test.py --verbose -i distributed/rpc/test_process_group_agent`. See internal test plan for internal details on testing with TCP init.

To avoid port collisions, we use `common.find_free_port()`.

Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Run RPC tests with TCP init in addition to file"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds testing with TCP init to the rpc tests. We disable/modify a few tests that would be broken under TCP initialization and will fix them in follow up PRs.

Running the tests with TCP init can be controlled with the env var setting `RPC_INIT_WITH_TCP="1"`. In OSS, we add a custom handler to run tests with the file init and TCP init in CI. OSS tests can be run with `python test/run_test.py --verbose -i distributed/rpc/test_process_group_agent`. See internal test plan for internal details on testing with TCP init.

To avoid port collisions, we use `common.find_free_port()`.

Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Run RPC tests with TCP init in addition to file"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds testing with TCP init to the rpc tests. We disable/modify a few tests that would be broken under TCP initialization and will fix them in follow up PRs.

Running the tests with TCP init can be controlled with the env var setting `RPC_INIT_WITH_TCP="1"`. In OSS, we add a custom handler to run tests with the file init and TCP init in CI. OSS tests can be run with `python test/run_test.py --verbose -i distributed/rpc/test_process_group_agent`. See internal test plan for internal details on testing with TCP init.

To avoid port collisions, we use `common.find_free_port()`.

Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

lw commented:
My main comment on this is whether we need to re-run the entire RPC suite to check for differences between file-based and TCP-based init. The suite already takes quite a while to run, and once we'll add GPU tests and tests for other backends (like shared memory) it might lead to a "combinatorial explosion" of the time it takes to run it.

Would there be a way to replicate what RPC does during its init, and add that as a test case for the Stores? I'll also take a look at the issue you referenced in the comments to get more context.

rohan-varma commented:
@lw Thanks for bringing that up, my line of thinking was that it could be worth it to run with TCP init as this change already caught ~4 tests that are broken with TCP init but fine with file init. 

We could always leave it as a flag and developers would have the option to set the flag to run with TCP init, but this probably doesn't eliminate the chance of checking in changes that may not work with TCP init. By running it in CI, there's probably a higher chance of preventing these bugs. If we're not sure about this, maybe we can just keep it as an env variable for now and not land the code that adds it as part of CI (at least in OSS, I think it's fine to run these internally)? 

For context around shared memory, that will be a transport built into TensorPipe right? Would we need to run all RPC tests under that transport or would it be sufficient to test that at the TensorPipe level?

rohan-varma committed:
Update on "Run RPC tests with TCP init in addition to file"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds testing with TCP init to the rpc tests. We disable/modify a few tests that would be broken under TCP initialization and will fix them in follow up PRs.

Running the tests with TCP init can be controlled with the env var setting `RPC_INIT_WITH_TCP="1"`. In OSS, we add a custom handler to run tests with the file init and TCP init in CI. OSS tests can be run with `python test/run_test.py --verbose -i distributed/rpc/test_process_group_agent`. See internal test plan for internal details on testing with TCP init.

To avoid port collisions, we use `common.find_free_port()`.

Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

lw commented:
> is the concern mainly from developer time/productivity perspective or CI env time/cost perspective (or both)?

I was mainly thinking in terms of time-to-signal for developers.

To ground this discussion in data, I looked at the CI logs of a recent job from master and from this PR:

- master
  https://app.circleci.com/pipelines/github/pytorch/pytorch/241692/workflows/63bedff3-95bb-46e3-9635-2382b5c4f3d5/jobs/9084940
  - ProcessGroup: from 10:29:10 to 10:40:51, hence ~11.5 minutes
  - TensorPipe: from 11:16:14 to 11:29:42, hence ~13.5 minutes

  Total: 25 minutes

- this PR
  https://app.circleci.com/pipelines/github/pytorch/pytorch/241655/workflows/c796c3dc-f85e-4308-882a-fde048b36122/jobs/9081850
  - ProcessGroup: from 06:02:32 to 06:26:48, hence ~24 minutes
  - TensorPipe: from 06:41:22 to 07:08:39, hence ~27 minutes

  Total: 51 minutes

It's just a single data point but it suggests that this PR would cause every PR to take 25 minutes longer to receive the green light from the CI. This may slow down developers (or encourage to land without waiting for CI).

Let me just clarify that I think we should absolutely run some tests with TCP. As you said, even just using it in our current tests revealed a bunch of errors that we should have caught earlier. However, those tests that fail do so because they're doing some "unusual" operations around initing and shutting down. My point is that tests of that type are the only ones that should be enrolled in TCP init, as they're the ones most likely to reveal something interesting. The vast majority of RPC tests do a "standard" init and then just test a wide variety of RPC-only features. It should be enough to run only one test of that type, as they all have effectively the same behavior when it comes to file-vs-TCP init.

Do you think we could take a pass over the RPC tests and try to select the ones we think would stress "unusual" code paths in init and tag them with a decorator, and then run only those with the TCP store?

> For context around shared memory, that will be a transport built into TensorPipe right? Would we need to run all RPC tests under that transport or would it be sufficient to test that at the TensorPipe level?

We didn't figure it out yet, but indeed if we were to run the entire RPC suite with the shared memory transports we'd also greatly increase the total PyTorch CI test time (by ~13.5 minutes, as only the TensorPipe suite needs to be repeated). This is certainly not great and ideally we should do as you say and test this transport in isolation within TensorPipe and then just "trust" that it works in PyTorch, without re-testing it. Unfortunately for now the PyTorch test suite is much more comprehensive than TensorPipe's own one... :/

rohan-varma commented:
> Do you think we could take a pass over the RPC tests and try to select the ones we think would stress "unusual" code paths in init and tag them with a decorator, and then run only those with the TCP store?

Sounds good, so I guess the proposal is the following:
- Don't run all tests with TCP init in addition to file in CI
- Add a decorator to run certain tests with the TCP init as well

rohan-varma committed:
Update on "Add an option to run RPC tests with TCP init"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds an environment variable `RPC_INIT_WITH_TCP` that allows us to run any RPC test with TCP initialization. 

To avoid port collisions, we use `common.find_free_port()`.
Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

rohan-varma commented:
@lw Updated the PR to just add the flag as an environment variable which will control whether the test runs with TCP or not, so that we can just run then under tcp init on an ad-hoc basis. I disabled a couple of the tests that don't work under it which we should fix.

lw reviewed (approved):
Ok, I'm fine with landing it as it is. It would be nice if we could enable some of these tests to run with TCP init in the CI at a later point.

rohan-varma committed:
Update on "Add an option to run RPC tests with TCP init"


We have found a few bugs where initializing/de-initializing/re-initializing RPC, and using RPC along with process groups does not work as expected, usually under TCP/env initialization (which is used over `file` which is the init that we use in our test in multi-machine scenarios).

Due to this motivation, this PR adds an environment variable `RPC_INIT_WITH_TCP` that allows us to run any RPC test with TCP initialization. 

To avoid port collisions, we use `common.find_free_port()`.
Differential Revision: [D25085458](https://our.internmc.facebook.com/intern/diff/D25085458/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25085458/)!

[ghstack-poisoned]

rohan-varma commented:
@lw Sounds good, there are 2 tests right now that probably should run with TCP init but don't work, I'll check and see if there are any additional.
